### ISSUE TITLE
A

### ISSUE THREAD

**Issue by soumith** [soumith/blog#1]
Intro: talk about the paper, overview etc.
Small-scale experiments: CIFAR stuff from Michael
   - Ablation studies
Large-scale experiments: Imagenet stuff from Sam
 - best practices found on the imagenet experiments
 - code optimizations to save memory and fit everything on GPUs
 - multi-GPU optimizations, talk about nccl etc.
Benchmarks
 - Just an inference / training time comparison vs VGG-A/D and Googlenet
Training Code
- Just briefly point to both training codes
Pre-trained models
- Declare that we are releasing pre-trained models, apart from training code.
- Showcase that the pre-trained models are higher accuracy than the original paper, attribute it to different data augmentation (maybe)
- Fine-tuning using the pre-trained models, point to the document on how one can use the pre-trained model for their own experiments etc.

**Comment by soumith** [comment]
Ablation studies

When trying to understand residual networks at first, it can be cumbersome to run exploratory studies on all of ImageNet since training a full model takes weeks to converge. Instead, it's often helpful to run *ablation studies* on a smaller dataset to independently measure the impact of each aspect of the model. This way, we can quickly understand what parts of the implementation are important to focus on during development. Quick turnaround time and continuous validation is helpful when designing the full system because small overlooked details can often bite you in the end.

For these experiments, we replicated Section 4.2 of the Residual Networks paper using the CIFAR dataset. In this setting, a small residual network with 20 layers takes about 8 hours to converge for 200 epochs on an EC2 g2.2xlarge instance and a larger 110-layer network takes 24 hours. This is still a long time, but critical bugs that prevent convergence are often immediately apparent because training loss should quickly begin making progress (within minutes).

**Effects of model depth.** The residual network paper provides a natural starting point for comparison: Figure 6 simply measures accuracy with respect to the depth of the network. To reproduce this figure, we held the learning rate policy and building block architecture fixed and varied the number of layers in the network between 20 and 118. Our results come fairly close to those in the paper: accuracy correlates well with model size, but levels off after 40 layers or so.

![CIFAR test error wrt. model depth](http://i.imgur.com/lt2D5cA.png)



**Building block architecture.** After validating that our results come fairly close to the original paper, we can consider the impact of slightly different building block architectures to test the assumptions of the model. For example:

- **Is it better to put batch normalization after the addition or before the addition at the end of each building block?** If batch normalization is placed after the addition, it has the effect of normalizing the output of the entire building block. This could be beneficial. However, this also forces every skip connection to perturb the output. This can be problematic: there are paths that allow data to pass through several successive batch normalization layers without any other processing. Each batch normalization layer applies its own separate distortion which compounds the original input. This has a harmful effect: we found that **putting batch normalization after the addition significantly hurts test error on CIFAR**, which is in line with the original paper’s recommendations.

- The above result seems to suggest that it's important to avoid changing data that passes only through identity connections. We can take this philosophy one step further: **should we remove the ReLU layers at the end of each building block?** ReLU layers also perturb data that flows through identity connections, but unlike batch normalization, ReLU's idempotence means that it doesn’t matter if data passes through one ReLU or thirty ReLUs. When we remove ReLU layers at the end of each building block, **we observe a small improvement in test performance** compared to the paper's suggested ReLU placement after the addition. However, the effect is fairly minor. More exploration is needed.

![Architecture testing error](http://i.imgur.com/sldN4cK.png)


These results were run on the deeper 110-layer models. The effect is much less pronounced on the shallow 20-layer baseline.

**Alternate solvers.** When running a hyperparameter search, it can often pay off to try fancier optimization strategies than vanilla SGD with momentum. Fancier solvers that make fancier assumptions may improve training times, but they may instead have more difficulty training these very deep models. In our experiments, we compared SGD+momentum (as used in the original paper) with RMSprop, Adadelta, and Adagrad. Many of them appear to converge faster initially (see the training curve below), but ultimately, SGD has 0.7% lower test error than the second-best strategy.

![Alternate solver training loss curve](http://i.imgur.com/0ZxQZ7k.png)



These experiments help verify the model’s correctness and uncover some interesting directions for future work. However, moving to the much larger ImageNet dataset opens its own pandora’s box of interesting challenges.

**Comment by soumith** [comment]
Could you post this to our google groups here: https://groups.google.com/forum/#!forum/torch7 , we only use github issues for bug tracking. We'll happily reply there. Thanks.
